crates/hc_sandbox/src/save.rs

```diff
@@ -2,215 +2,430 @@
+            .write(true)
+            .truncate(true)
+            .create(true)
+            .open(hc_file)?;
+        for path in self.existing_all.iter() {
+            match path {
+                Ok(path) => writeln!(file, "{}", path.display())?,
+                Err(path) => writeln!(file, "{}", path.display())?,
             }
         }
-        hc_dir.push(".hc");
-        if hc_dir.exists() {
-            std::fs::remove_file(&hc_dir)
-                .with_context(|| format!("Failed to remove .hc file at {}", hc_dir.display()))?;
-        }
-        hc_dir.pop();
-        hc_dir.push(".hc_auth");
-        if hc_dir.exists() {
-            std::fs::remove_file(&hc_dir).with_context(|| {
-                format!("Failed to remove .hc_auth file at {}", hc_dir.display())
-            })?;
-        }
+
+        Ok(())
     }
-    Ok(())
-}
 
-/// Load sandbox paths from the `.hc` file.
-pub fn load(mut hc_dir: PathBuf) -> anyhow::Result<Vec<PathBuf>> {
-    let mut paths = Vec::new();
-    hc_dir.push(".hc");
-    if hc_dir.exists() {
-        let existing = std::fs::read_to_string(hc_dir)?;
-        for sandbox in existing.lines() {
-            let path = PathBuf::from(sandbox);
-            let config_file_path = ConfigFilePath::from(ConfigRootPath::from(path.clone()));
-            if config_file_path.as_ref().exists() {
-                paths.push(path);
+    /// Append paths to HcFile and save to disk
+    pub fn append(&mut self, paths: Vec<ConfigRootPath>) -> std::io::Result<()> {
+        for path in paths.into_iter() {
+            if path.exists() && path.is_dir() {
+                self.existing_all.push(Ok(path));
             } else {
-                tracing::error!("Failed to load path {} from existing .hc", path.display());
+                self.existing_all.push(Err(path));
             }
         }
+        self.save()
     }
-    Ok(paths)
-}
 
-/// Print out the sandboxes contained in the `.hc` file.
-pub fn list(hc_dir: PathBuf, verbose: bool) -> anyhow::Result<()> {
-    let out = load(hc_dir)?.into_iter().enumerate().try_fold(
-        "\nSandboxes contained in `.hc`\n".to_string(),
-        |out, (i, path)| {
-            let r = match verbose {
-                false => format!("{}{}: {}\n", out, i, path.display()),
-                true => {
-                    let config = holochain_conductor_config::config::read_config(
-                        ConfigRootPath::from(path.clone()),
-                    )?;
+    /// Remove paths by their index in the `.hc` file
+    /// and attempt to delete the sandbox folder.
+    /// If no indices are passed in then they will all be removed.
+    /// If no sandbox paths remain then all `.hc_*` files will be removed.
+    pub fn remove(mut self, indices_to_remove: Vec<usize>) -> std::io::Result<usize> {
+        let cur_size = self.existing_all.len();
+        let mut to_remove = Vec::new();
+        let mut remaining = Vec::new();
+        if indices_to_remove.is_empty() {
```

Comment:
This is a then clean operation. I get that the CLI protects the user by this is library code too and this would be really easy to get wrong.